### PR TITLE
[WIP] Annotate DataFrame (Part 3)

### PR DESCRIPTION
**Pull request by vaibhavhrt**
- [x] part of #26792 
- [ ] tests added / passed
- [x] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -994,13 +1004,13 @@ def itertuples(self, index=True, name="Pandas"):
...
 
-    def __len__(self):
+    def __len__(self) -> int:
         """
         Returns length of info axis, but here we use the index.
         """
         return len(self.index)
 
-    def dot(self, other):
+    def dot(self, other: Union[Series, DataFrame]) -> Union[Series, DataFrame]:
```
Comment (WillAyd) [inline pandas/core/frame.py]: Should include `np.ndarray` in the signature here. Applicable to the rest below as well

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 1 above)
Comment (vaibhavhrt) [reply pandas/core/frame.py]: Alright.
Added some types without paying much attention, will take a closer later.
Btw I am guessing if type of `other` is `Series`, return type will be `Series` too, similar for `DataFrame` and `np.ndarray`, how do I indicate that.

Comment (WillAyd) [reply pandas/core/frame.py]: A TypeVar would make the function generic (i.e. the return type would match `other` if the same TypeVar is used for both) but unfortunately I don't think that is how this is actually implemented

File: pandas/core/frame.py [hunk 2]
```diff
@@ -850,7 +860,7 @@ def style(self):
         """
 
     @Appender(_shared_docs["items"])
-    def items(self):
+    def items(self) -> Generator[Tuple[str, Dtype], None, None]:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py]: A basic generator that only yields values can be annotated as having a return type of either Iterator[YieldType] or Iterable[YieldType]. https://mypy.readthedocs.io/en/latest/kinds_of_types.html#generators

my preference would be Iterable.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -862,7 +872,7 @@ def items(self):
     def iteritems(self):
         yield from self.items()
 
-    def iterrows(self):
+    def iterrows(self) -> Generator[Tuple[Index, Series], None, None]:
```
Comment (simonjayhawkins) [inline pandas/core/frame.py]: same comment as above.

File: pandas/core/frame.py (hunk 1 above)
Comment (vaibhavhrt) [reply pandas/core/frame.py]: hmm, pretty sure saw something like this in the docstring. But did this in a hurry(thats why draft PR), a bit busy. Will take a closer look as soon as I can.

**Comment by simonjayhawkins** [comment]
closed as stale. @vaibhavhrt ping if you wish to continue.

**Comment by vaibhavhrt** [comment]
Sorry been really busy last few weeks. Will be busy next few days as well,
but I do wish to continue and complete it, just give me a few days I will
reopen the PR.

File: pandas/core/frame.py [hunk 4]
```diff
@@ -4825,7 +4827,7 @@ def drop_duplicates(self, subset=None, keep="first", inplace=False):
         duplicated = self.duplicated(subset, keep=keep)
 
         if inplace:
-            inds, = (-duplicated)._ndarray_values.nonzero()
+            (inds,) = (-duplicated)._ndarray_values.nonzero()
```
Comment (jbrockmendel) [inline pandas/core/frame.py]: pls use `inds = ...[0]`

File: pandas/core/frame.py [hunk 5]
```diff
@@ -1022,13 +1022,15 @@ def itertuples(self, index=True, name="Pandas"):
...
+    def __len__(self) -> int:
         """
         Returns length of info axis, but here we use the index.
         """
         return len(self.index)
 
-    def dot(self, other):
+    def dot(
+        self, other: Union[Series, DataFrame, np.ndarray]
+    ) -> Union[Series, DataFrame, np.ndarray]:
```
Comment (jbrockmendel) [inline pandas/core/frame.py]: can we get ndarray back?  id expect it to always be Series or DataFrame

File: pandas/core/frame.py [hunk 6]
```diff
@@ -812,7 +812,7 @@ def to_string(
     # ----------------------------------------------------------------------
 
     @property
-    def style(self):
+    def style(self) -> "Styler":
```
Comment (jbrockmendel) [inline pandas/core/frame.py]: mypy complains about this, needs Styler imported in the TYPE_CHECKING block

File: pandas/core/frame.py (hunk 4 above)
Comment (vaibhavhrt) [reply pandas/core/frame.py]: done

File: pandas/core/frame.py (hunk 6 above)
Comment (vaibhavhrt) [reply pandas/core/frame.py]: sorry not sure what you mean, can you please explain a bit

File: pandas/core/frame.py (hunk 5 above)
Comment (vaibhavhrt) [reply pandas/core/frame.py]: removed ndarray

**Comment by vaibhavhrt** [comment]
removing from draft, otherwise I keep forgetting about it. Will try it get it done fast now.

File: pandas/core/frame.py (hunk 6 above)
Comment (vaibhavhrt) [reply pandas/core/frame.py]: nvm, fixed figured it out.

File: pandas/_typing.py [hunk 7]
```diff
@@ -23,6 +23,7 @@
     from pandas.core.indexes.base import Index  # noqa: F401
     from pandas.core.series import Series  # noqa: F401
     from pandas.core.generic import NDFrame  # noqa: F401
+    from pandas.io.formats.style import Styler  # noqa: F401
```
Comment (WillAyd) [inline pandas/_typing.py]: Just put this in a `if TYPE_CHECKING` block in pandas.core.frame instead of here

File: pandas/core/frame.py [hunk 8]
```diff
@@ -4631,7 +4631,7 @@ def drop_duplicates(self, subset=None, keep="first", inplace=False):
         duplicated = self.duplicated(subset, keep=keep)
 
         if inplace:
-            (inds,) = (-duplicated)._ndarray_values.nonzero()
+            inds = (-duplicated)._ndarray_values.nonzero()[0]
```
Comment (WillAyd) [inline pandas/core/frame.py]: Was this required for mypy? If not can you revert?

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
closing as stale, ping if you want to continue.

**Comment by stangirala** [comment]
@WillAyd are we working on this (and https://github.com/pandas-dev/pandas/issues/26792)?

**Comment by stangirala** [comment]
I can help out here, I really want to see type checking with a df.

**Comment by WillAyd** [comment]
@stangirala this PR specifically is stale; if you have contributions you'd like to make for that feel free to push a separate PR

**Comment by stangirala** [comment]
@WillAyd can I use this PR as a template?

**Pull request closed without merging**